### DISCUSSION TITLE
CRUD Operations on namespaces

### DISCUSSION THREAD

**Discussion by dialalpha** [oxia-db/oxia#445] [Q&A] [answered]
IIUC all operations on keys occur within a specific namespace. However I don't see any API that allows listing namespaces for example. Is this by design or something that will be added later?

**Comment by merlimat** [comment, marked as answer]
We had some thoughts about adding an "admin" API, though nothing concrete yet. 

The namespaces are controlled through the K8S config map (or operator CRD). Once you add the namespace to the config map, the coordinator will make sure the shards for the namespaces are created and available. Similarly, once you remove the namespace from the config map, the shards and the data will be deleted.

One thing to keep in mind is that Oxia namespaces do not map to Pulsar namespaces. In general, Pulsar would use just a handful of namespaces. eg: pulsar, bookkeeper, functions.

**Reply by dialalpha** [reply]
Thanks for the clarification @merlimat. Does this mean that oxia namespaces are fixed once the cluster is running?

**Reply by merlimat** [reply]
No, when the config map is updated, the coordinator will automatically pick-up the change and add/remove the shards for the namespace.

**Comment by dialalpha** [comment]
Do you have any concerns that the 1MB limit on the size of the ConfigMap might be too low? Say if you had a use case where you want to share the oxia cluster among multiple tenants, each requiring multiple namespaces.

**Reply by merlimat** [reply]
From the pov of the ConfigMap, each namespace would probably take <100 bytes. That gives for ~10K namespaces. 

We would encounter other practical problems before that: 
 1. The cluster status is also kept in a single ConfigMap, though the size per namespace would be slightly bigger (eg: leader election info per each shard). 
 2. Having 10K separate DBs (considering 1 shard per namespace)  would also require quite a bit of resources. 
 
For (1) and for the namespaces configMap, we can find a better solution (eg: split into multiple config maps).

For (2) it becomes a bit trickier. To support it well, we should part ways with the current implementation of keeping 1 DB per shard, and instead having 1 DB per storage server with per-shard key prefixes. 

In short: the initial assumption for namespaces has been to have ~ 100-1k max namespaces in an Oxia cluster. Supporting more namespace would definitely be possible, though would require some changes.

**Reply by dialalpha** [reply]
Thanks @merlimat. That was very informative.
